itou/www/search/views.py

```diff
@@ -27,3 +33,30 @@ def search_siaes(request, template_name="search/siaes_search_results.html"):
 
     context = {"form": form, "siaes_page": siaes_page}
     return render(request, template_name, context)
+
+
+def get_shuffled_rank():
+    """
+    Quick and dirty solution to shuffle results with a
+    determistic seed which changes every day.
+
+    We may later implement a more rigorous shuffling but this will
+    need an extra column in db.
+
+    Note that we have about 3K siaes.
+
+    We produce a large pseudo-random integer on the fly from id
+    with the static PG expression `(A+id)*(B+id)`
+    which has the advantage of still be using index scan by PG.
+    It is important that this large integer is far from zero to avoid
+    that id=1,2,3 always stay on the top of the list.
+    Thus we choose rather large A and B.
+
+    We then take a modulo which changes everyday.
+    """
+    # Seed changes every day at midnight.
+    random.seed(datetime.date.today())
+    a = random.randint(1000, 10000)
+    b = random.randint(1000, 10000)
+    modulo = random.randint(100, 1000)
+    return (a + F("id")) * (b + F("id")) % modulo
```

Comment:
Ok, ça y est j'ai compris ! Ça monte pas au cerveau tout de suite mais ça marche bien. Je connaissais pas `random.seed`.

J'ai pas mieux à proposer.

Par contre, ton commentaire me semble faux puisqu'il est impossible d'utiliser un index avec un `annotate` généré dynamiquement (et donc non indexable préalablement).

Cela dit c'est pas pire qu'avant ! Et en plus j'ai une *loop* à *queries* dégueulasse dans le template qu'il faudra fixer à l'occasion :D

Juste en termes d'organisation du code j'aurais fait de `.annotate(shuffled_rank=get_shuffled_rank())` une méthode de `SiaeQuerySet` pour ne pas mettre ce code dans un fichier `view`.

Reply:
> Juste en termes d'organisation du code j'aurais fait de .annotate(shuffled_rank=get_shuffled_rank()) une méthode de SiaeQuerySet pour ne pas mettre ce code dans un fichier view.

done

Reply:
> Et en plus j'ai une loop à queries dégueulasse dans le template qu'il faudra fixer à l'occasion

J'ai regardé un peu, c'est le `{% if siae.has_members %}` qui est appelé 10 fois (une fois par siae) et fait donc 10 queries PG. Pourtant on fait un `prefetch_related("members")` donc tous les members y-compris leur champ `is_active` sont normalement loadés, du coup `siae.has_members` qui s'appuie sur `siae.active_members` soit `self.members.filter(is_active=True)` devrait fonctionner sans query supplémentaire.

Je ne comprends pas tout sur ce cas, donc pas de quick fix pour cette fois sorry.

Reply:
> Par contre, ton commentaire me semble faux puisqu'il est impossible d'utiliser un index avec un annotate généré dynamiquement (et donc non indexable préalablement).

AFAIU L'index sera utilisé pour toute expression statique basée uniquement sur une colonne indexée sans autre variable. Du point de vue de PG `((a+id)*(b+id)) % c` est une simple fonction de `id` tout comme le serait `id + 42` par exemple. Car a,b,c sont des constantes.

Et mes observations du EXPLAIN le confirment.

Reply:
Pas de soucis, je fixerai asap !

Reply:
re @kemar 

MER/MEP pour bientôt

Reply:
Ok je ne suis pas tout à fait certain que ton explication sur l'indexation soit correcte mais je creuserai à l'occasion. En tous cas c'est nickel comme solution.

Reply:
First comment ever!

Reply:
> AFAIU L'index sera utilisé pour toute expression statique basée uniquement sur une colonne indexée sans autre variable. Du point de vue de PG `((a+id)*(b+id)) % c` est une simple fonction de `id` tout comme le serait `id + 42` par exemple. Car a,b,c sont des constantes.
> 
> Et mes observations du EXPLAIN le confirment.

Je ne sais pas comment tu fais ton EXPLAIN mais pour cette *query*…

```sql
EXPLAIN ANALYSE SELECT id, (((8385 + id) * (1551 + id)) % 7577) AS "shuffled_rank"
FROM siaes_siae
WHERE department IN ('62', '67', '93')
ORDER BY "shuffled_rank" ASC
```

…voici le résultat :

> https://explain.depesz.com/s/8KOe

S'il y a bien un index scan sur le département, c'est un heap scan sur `shuffled_rank` et il ne peut pas en être autrement puisque pour utiliser un index ce dernier doit exister au préalable.

Reply:
J'ai manifestement vu un fantôme... globalement j'étais tellement focus sur l'absence de `Seq Scan` et satisfait du `Bitmap Index Scan` pour le département, qu'ensuite en passant sur le shuffling j'ai vu `Bitmap * Scan` et je me suis dit "parfait c'est pas un seq scan donc c'est bon aussi", sans me rendre compte que c'était un Heap et non Index scan o_O

Second fantôme, ma fixette sur ce commentaire:

![image](https://user-images.githubusercontent.com/10533583/77167811-63918380-6abf-11ea-9065-8e25df8c8392.png)
 
En fait pour avoir le Index Scan sur `((8385 + id) * (1551 + id)) % 7577` il faut manuellement ajouter un index sur cette expression statique (TIL on peut faire un index sur une expression, et non pas juste sur une colonne ou une combinaison ordonnée de colonnes, I had no idea).

Tout n'est pas si noir, le Bitmap Heap Scan est quand même "un peu mieux" qu'un Seq Scan mais je reconnais que c'est pas top du tout.

J'ai relu http://media.pragprog.com/titles/bksqla/random.pdf avec un nouveau oeil, et notre meilleur espoir est peut-être la section `Choose a Random Key Value Between 1 and MAX - One technique that avoids sorting the table is to choose a random value
between 1 and the greatest primary key value.` car on a la chance que nos IDs de siaes commencent à 1 et sont presque sans trous.

En couplant cela avec un https://www.techonthenet.com/postgresql/functions/setseed.php pour changer le seed chaque soir à minuit, je peux peut-être y arriver.

Je creuse cela cet après-midi.

Reply:
En passant, excellent le https://explain.depesz.com/ ! TIL.

Reply:
Ça n'est pas noir du tout tu as raison, et ça fait largement l'affaire et de façon élégante. Je voulais juste que ça soit bien clair que ça ne pouvait pas utiliser un index.

Reply:
Bon en fait je vais nulle part... la technique que je regardais dans http://media.pragprog.com/titles/bksqla/random.pdf permet seulement de fetcher 1 row au hasard, pas de mélanger les rows en utilisant uniquement du index scan.

A ce stade, la seule façon que je vois de faire tout cela uniquement en index scan, c'est de rajouter un index sur l'expression `((8385 + id) * (1551 + id)) % 7577` et de faire un cronjob chaque soir à minuit qui reconstruit ce même index avec des valeurs différentes des trois constantes. Et à ce moment là on oublie même tout la plomberie du modulo&co, un bête random différent chaque soir suffira.

Je pense qu'on y échappera pas si on veut vraiment du index scan.

Un argument médiocre mais pour se consoler: la table siaes devrait rester dans les 3K-10K rows à terme. C'est surtout la table des candidatures par exemple qui devrait nous faire souffrir en premier avec son volume explosif à terme.

Et l'autre argument médiocre pour se consoler aussi: du bitmap heap scan, c'est quand même un peu mieux que du seq scan.

J'arrête là pour ça pour aujourd'hui.

A tes bons commentaires @kemar

Reply:
Fixed doc about this. https://github.com/betagouv/itou/commit/2d7203675e45b49eb1aed54088df268900b8fca9

Reply:
Oui je suis d'accord c'est chaud d'indexer ça ! La solution que tu as mise en place est bonne.

Je voulais juste que tu fixe ton commentaire, ce que tu as fait, donc merci :)